pytorch_lightning/utilities/device_parser.py

```diff
@@ -241,10 +240,42 @@ def _check_data_type(device_ids: Any) -> None:
         MisconfigurationException:
             If ``device_ids`` of GPU/TPUs aren't ``int``, ``str``, sequence of ``int`` or ``None``
     """
-    if device_ids is not None and (
-        not isinstance(device_ids, (int, str, MutableSequence, tuple)) or isinstance(device_ids, bool)
-    ):
-        raise MisconfigurationException("Device ID's (GPU/TPU) must be int, string or sequence of ints or None.")
+    # A flag which is set when incompatible input is passed
+    raise_error_flag = False
+    # If device_ids is an iterator
+    is_iterator = isinstance(device_ids, (MutableSequence, tuple))
+    # If an iterator, this will contain the type inside the iterator, else the type of device_ids
+    incompatible_type = None
+
+    if device_ids is None or (is_iterator and len(device_ids) == 0):
+        return
+
+    if (not is_iterator and not isinstance(device_ids, (int, str))) or isinstance(device_ids, bool):
+        raise_error_flag = True
+        incompatible_type = type(device_ids).__name__
+
+    if is_iterator:
+        # Iterate through the sequence, and assert it's 1 dimensional and only contains integers
+        for _iter in iter(device_ids):
+            if not isinstance(_iter, int):
```

Comment:
can't we do `all(isinstance(x, int) for x in device_ids)`

Reply:
Good question, @awaelchli! We can, but I also want to get the type of x in `device_ids` if it's not an integer. Hence this loop.